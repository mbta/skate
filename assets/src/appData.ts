const appData = (): DOMStringMap | undefined => {
  const dataEl = document.getElementById("app")

  if (!dataEl) {
    return undefined
  }

  return dataEl.dataset
}

export default appData
