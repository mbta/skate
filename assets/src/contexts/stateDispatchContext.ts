import { createContext } from "react"
import { Dispatch, initialState, State } from "../state"

type StateDispatchContextData = [State, Dispatch]

// tslint:disable-next-line: no-empty
const noopDispatch: Dispatch = () => {}
const StateDispatchContext = createContext([
  initialState,
  noopDispatch,
] as StateDispatchContextData)

export default StateDispatchContext
