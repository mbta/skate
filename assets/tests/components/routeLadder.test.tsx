import { mount } from "enzyme"
import React from "react"
import renderer, { act } from "react-test-renderer"
import { LadderDirection } from "../../src/components/ladder"
import RouteLadder, { groupByPosition } from "../../src/components/routeLadder"
import { StateDispatchProvider } from "../../src/contexts/stateDispatchContext"
import { HeadwaySpacing } from "../../src/models/vehicleStatus"
import {
  Ghost,
  RouteStatus,
  Vehicle,
  VehicleOrGhost,
} from "../../src/realtime.d"
import { Route } from "../../src/schedule.d"
import { deselectRoute, initialState, selectVehicle } from "../../src/state"

// tslint:disable: object-literal-sort-keys

jest.mock("../../src/laboratoryFeatures", () => ({
  __esModule: true,
  default: () => true,
}))

const vehicles: Vehicle[] = [
  {
    id: "y1818",
    label: "1818",
    runId: "run-1",
    timestamp: 1557160307,
    latitude: 0,
    longitude: 0,
    directionId: 0,
    routeId: "28",
    tripId: "39914237",
    operatorId: "op1",
    operatorName: "SMITH",
    bearing: 33,
    blockId: "block-1",
    headsign: "h1",
    viaVariant: "4",
    headwaySecs: 859.1,
    headwaySpacing: HeadwaySpacing.Ok,
    previousVehicleId: "v2",
    scheduleAdherenceSecs: 0,
    scheduledHeadwaySecs: 120,
    isOffCourse: false,
    layoverDepartureTime: null,
    blockIsActive: true,
    dataDiscrepancies: [],
    stopStatus: {
      stopId: "57",
      stopName: "57",
    },
    timepointStatus: {
      fractionUntilTimepoint: 0.5,
      timepointId: "MATPN",
    },
    scheduledLocation: null,
    routeStatus: "on_route",
  },
  {
    id: "y0479",
    label: "0479",
    runId: "run-2",
    timestamp: 1557160347,
    latitude: 0,
    longitude: 0,
    directionId: 1,
    routeId: "28",
    tripId: "39914128",
    operatorId: "op2",
    operatorName: "JONES",
    bearing: 33,
    blockId: "block-1",
    headsign: null,
    viaVariant: null,
    headwaySecs: 859.1,
    headwaySpacing: HeadwaySpacing.Ok,
    previousVehicleId: "v2",
    scheduleAdherenceSecs: 0,
    scheduledHeadwaySecs: 120,
    isOffCourse: false,
    layoverDepartureTime: null,
    blockIsActive: true,
    dataDiscrepancies: [],
    stopStatus: {
      stopId: "59",
      stopName: "59",
    },
    timepointStatus: {
      fractionUntilTimepoint: 0.0,
      timepointId: "MORTN",
    },
    scheduledLocation: {
      directionId: 1,
      timepointStatus: {
        timepointId: "MORTN",
        fractionUntilTimepoint: 0.0,
      },
    },
    routeStatus: "on_route",
  },
]

describe("routeLadder", () => {
  const originalGetBBox = SVGSVGElement.prototype.getBBox
  const originalGetElementsByClassName = document.getElementsByClassName

  beforeEach(() => {
    SVGSVGElement.prototype.getBBox = () => {
      return {
        height: 100,
        width: 50,
      } as DOMRect
    }
    const mockElement = {
      offsetHeight: 120,
    }
    // @ts-ignore
    document.getElementsByClassName = () => [mockElement]
  })

  afterEach(() => {
    SVGSVGElement.prototype.getBBox = originalGetBBox
    document.getElementsByClassName = originalGetElementsByClassName
  })

  test("renders a route ladder", () => {
    const route: Route = {
      id: "28",
      directionNames: { 0: "Outbound", 1: "Inbound" },
      name: "28",
    }
    const timepoints = ["MATPN", "WELLH", "MORTN"]

    const tree = renderer
      .create(
        <RouteLadder
          route={route}
          timepoints={timepoints}
          vehiclesAndGhosts={undefined}
          selectedVehicleId={undefined}
        />
      )
      .toJSON()

    expect(tree).toMatchSnapshot()
  })

  test("renders a route ladder with vehicles", () => {
    const route: Route = {
      id: "28",
      directionNames: { 0: "Outbound", 1: "Inbound" },
      name: "28",
    }
    const timepoints = ["MATPN", "WELLH", "MORTN"]

    const ghost: Ghost = {
      id: "ghost-trip",
      directionId: 0,
      routeId: route.id,
      tripId: "ghost trip",
      headsign: "headsign",
      blockId: "ghost block",
      runId: "123-0123",
      viaVariant: null,
      layoverDepartureTime: null,
      scheduledTimepointStatus: {
        timepointId: "MORTN",
        fractionUntilTimepoint: 0.0,
      },
      routeStatus: "on_route",
    }

    const tree = renderer
      .create(
        <RouteLadder
          route={route}
          timepoints={timepoints}
          vehiclesAndGhosts={(vehicles as VehicleOrGhost[]).concat([ghost])}
          selectedVehicleId={undefined}
        />
      )
      .toJSON()

    expect(tree).toMatchSnapshot()
  })

  test("renders a route ladder with vehicles in the incoming box", () => {
    const route: Route = {
      id: "28",
      directionNames: { 0: "Outbound", 1: "Inbound" },
      name: "28",
    }
    const timepoints = ["MATPN", "WELLH", "MORTN"]
    const tree = renderer
      .create(
        <RouteLadder
          route={route}
          timepoints={timepoints}
          vehiclesAndGhosts={vehicles.map((vehicle: Vehicle) => ({
            ...vehicle,
            routeStatus: "pulling_out" as RouteStatus,
          }))}
          selectedVehicleId={undefined}
        />
      )
      .toJSON()

    expect(tree).toMatchSnapshot()
  })

  test("renders a route ladder with laying over vehicles", () => {
    const route: Route = {
      id: "28",
      directionNames: { 0: "Outbound", 1: "Inbound" },
      name: "28",
    }

    const timepoints = ["MATPN", "WELLH", "MORTN"]

    const [v1, v2] = vehicles
    const tree = renderer
      .create(
        <RouteLadder
          route={route}
          selectedVehicleId={undefined}
          timepoints={timepoints}
          vehiclesAndGhosts={[
            { ...v1, routeStatus: "laying_over" },
            { ...v2, routeStatus: "laying_over" },
          ]}
        />
      )
      .toJSON()

    expect(tree).toMatchSnapshot()
  })

  test("displays loading if we are fetching the timepoints", () => {
    const route: Route = {
      id: "28",
      directionNames: { 0: "Outbound", 1: "Inbound" },
      name: "28",
    }
    const timepoints = null

    const tree = renderer
      .create(
        <RouteLadder
          route={route}
          timepoints={timepoints}
          vehiclesAndGhosts={undefined}
          selectedVehicleId={undefined}
        />
      )
      .toJSON()

    expect(tree).toMatchSnapshot()
  })

  test("clicking the close button deselects that route", () => {
    const mockDispatch = jest.fn()
    const route: Route = {
      id: "28",
      directionNames: { 0: "Outbound", 1: "Inbound" },
      name: "28",
    }
    const timepoints = ["MATPN", "WELLH", "MORTN"]

    const wrapper = mount(
      <StateDispatchProvider state={initialState} dispatch={mockDispatch}>
        <RouteLadder
          route={route}
          timepoints={timepoints}
          vehiclesAndGhosts={undefined}
          selectedVehicleId={undefined}
        />
      </StateDispatchProvider>
    )
    wrapper.find(".m-route-ladder__header .m-close-button").simulate("click")

    expect(mockDispatch).toHaveBeenCalledWith(deselectRoute("28"))
  })

  test("clicking the reverse button reverses the order of the timepoints", () => {
    const route: Route = {
      id: "28",
      directionNames: { 0: "Outbound", 1: "Inbound" },
      name: "28",
    }
    const timepoints = ["MATPN", "WELLH", "MORTN"]

    const wrapper = mount(
      <RouteLadder
        route={route}
        timepoints={timepoints}
        vehiclesAndGhosts={undefined}
        selectedVehicleId={undefined}
      />
    )
    act(() => {
      wrapper.find(".m-route-ladder__reverse").simulate("click")
    })

    expect(
      wrapper.find(".m-ladder__timepoint-name").map(node => node.text())
    ).toEqual(["MORTN", "WELLH", "MATPN"])
  })

  test("clicking an incoming vehicle selects that vehicle", () => {
    const mockDispatch = jest.fn()

    const route: Route = {
      id: "28",
      directionNames: { 0: "Outbound", 1: "Inbound" },
      name: "28",
    }
    const timepoints = ["MATPN", "WELLH", "MORTN"]
    const vehicle: Vehicle = {
      id: "v1",
      label: "v1",
      runId: "run1",
      timestamp: 0,
      latitude: 0,
      longitude: 0,
      directionId: 1,
      routeId: "28",
      tripId: "trip",
      headsign: null,
      viaVariant: null,
      operatorId: "op1",
      operatorName: "SMITH",
      bearing: 33,
      blockId: "block-1",
      headwaySecs: 859.1,
      headwaySpacing: HeadwaySpacing.Ok,
      previousVehicleId: "v2",
      scheduleAdherenceSecs: 0,
      scheduledHeadwaySecs: 120,
      isOffCourse: false,
      layoverDepartureTime: null,
      blockIsActive: true,
      dataDiscrepancies: [],
      stopStatus: {
        stopId: "stop",
        stopName: "stop",
      },
      timepointStatus: {
        timepointId: "MATPN",
        fractionUntilTimepoint: 0.5,
      },
      scheduledLocation: null,
      routeStatus: "pulling_out",
    }

    const wrapper = mount(
      <StateDispatchProvider state={initialState} dispatch={mockDispatch}>
        <RouteLadder
          route={route}
          timepoints={timepoints}
          vehiclesAndGhosts={[vehicle]}
          selectedVehicleId={undefined}
        />
      </StateDispatchProvider>
    )
    wrapper.find(".m-incoming-box__vehicle").simulate("click")

    expect(mockDispatch).toHaveBeenCalledWith(selectVehicle(vehicle.id))
  })
})

describe("groupByPosition", () => {
  const emptyByPosition = {
    onRoute: [],
    layingOverTop: [],
    layingOverBottom: [],
    incoming: [],
  }
  test("loading", () => {
    expect(groupByPosition(undefined, "1", LadderDirection.ZeroToOne)).toEqual(
      emptyByPosition
    )
  })

  test("on route", () => {
    const vehicle: Vehicle = {
      id: "y0001",
      routeId: "1",
      directionId: 0,
      routeStatus: "on_route",
    } as Vehicle
    expect(groupByPosition([vehicle], "1", LadderDirection.ZeroToOne)).toEqual({
      ...emptyByPosition,
      onRoute: [vehicle],
    })
  })

  test("laying over", () => {
    const ladderDirection: LadderDirection = LadderDirection.ZeroToOne
    const top: Vehicle = {
      id: "y0001",
      routeId: "1",
      directionId: 1,
      routeStatus: "laying_over",
    } as Vehicle
    const bottom: Vehicle = {
      id: "y0002",
      routeId: "1",
      directionId: 0,
      routeStatus: "laying_over",
    } as Vehicle
    expect(groupByPosition([top, bottom], "1", ladderDirection)).toEqual({
      ...emptyByPosition,
      layingOverTop: [top],
      layingOverBottom: [bottom],
    })
  })

  test("pulling out", () => {
    const vehicle: Vehicle = {
      id: "y0001",
      routeId: "1",
      directionId: 0,
      routeStatus: "pulling_out",
    } as Vehicle
    expect(groupByPosition([vehicle], "1", LadderDirection.ZeroToOne)).toEqual({
      ...emptyByPosition,
      incoming: [vehicle],
    })
  })

  test("incoming from another route", () => {
    const vehicle: Vehicle = {
      id: "y0001",
      routeId: "2",
      directionId: 0,
      routeStatus: "on_route",
    } as Vehicle
    expect(groupByPosition([vehicle], "1", LadderDirection.ZeroToOne)).toEqual({
      ...emptyByPosition,
      incoming: [vehicle],
    })
  })

  test("on route ghost", () => {
    const ghost: Ghost = {
      id: "ghost",
      routeId: "1",
      directionId: 0,
      routeStatus: "on_route",
    } as Ghost
    expect(groupByPosition([ghost], "1", LadderDirection.ZeroToOne)).toEqual({
      ...emptyByPosition,
      onRoute: [ghost],
    })
  })
})
