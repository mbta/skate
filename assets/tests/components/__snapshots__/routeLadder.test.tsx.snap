// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`routeLadder displays loading if we are fetching the timepoints 1`] = `
Array [
  <div
    className="m-route-ladder__header"
  >
    <button
      className="m-close-button"
      onClick={[Function]}
    >
      <span
        className=""
        dangerouslySetInnerHTML={
          Object {
            "__html": "SVG",
          }
        }
      />
    </button>
    <div
      className="m-route-ladder__route-name"
    >
      28
    </div>
  </div>,
  <div
    className="m-route-ladder__controls"
  >
    <button
      className="m-route-ladder__reverse"
      onClick={[Function]}
    >
      <span
        className="m-route-ladder__reverse-icon"
        dangerouslySetInnerHTML={
          Object {
            "__html": "SVG",
          }
        }
      />
      Reverse
    </button>
  </div>,
  "loading...",
]
`;

exports[`routeLadder renders a route ladder 1`] = `
Array [
  <div
    className="m-route-ladder__header"
  >
    <button
      className="m-close-button"
      onClick={[Function]}
    >
      <span
        className=""
        dangerouslySetInnerHTML={
          Object {
            "__html": "SVG",
          }
        }
      />
    </button>
    <div
      className="m-route-ladder__route-name"
    >
      28
    </div>
  </div>,
  <div
    className="m-route-ladder__controls"
  >
    <button
      className="m-route-ladder__reverse"
      onClick={[Function]}
    >
      <span
        className="m-route-ladder__reverse-icon"
        dangerouslySetInnerHTML={
          Object {
            "__html": "SVG",
          }
        }
      />
      Reverse
    </button>
  </div>,
  <div
    className="m-layover-box m-layover-box--top"
  />,
  <div
    className="m-ladder"
    style={
      Object {
        "width": 184,
      }
    }
  >
    <svg
      className="m-ladder__svg"
      height={0}
      viewBox="-92 -20 184 0"
      width={184}
    >
      <line
        className="m-ladder__line"
        x1={-40}
        x2={-40}
        y1="0"
        y2={-40}
      />
      <line
        className="m-ladder__line"
        x1={40}
        x2={40}
        y1="0"
        y2={-40}
      />
      <circle
        className="m-ladder__stop-circle"
        cx={-40}
        cy={-0}
        r="3"
      />
      <circle
        className="m-ladder__stop-circle"
        cx={40}
        cy={-0}
        r="3"
      />
      <text
        className="m-ladder__timepoint-name"
        data-event="click"
        data-tip="MATPN Name"
        dominantBaseline="middle"
        textAnchor="middle"
        x="0"
        y={-0}
      >
        MATPN
      </text>
      <circle
        className="m-ladder__stop-circle"
        cx={-40}
        cy={-20}
        r="3"
      />
      <circle
        className="m-ladder__stop-circle"
        cx={40}
        cy={-20}
        r="3"
      />
      <text
        className="m-ladder__timepoint-name"
        data-event="click"
        data-tip="WELLH Name"
        dominantBaseline="middle"
        textAnchor="middle"
        x="0"
        y={-20}
      >
        WELLH
      </text>
      <circle
        className="m-ladder__stop-circle"
        cx={-40}
        cy={-40}
        r="3"
      />
      <circle
        className="m-ladder__stop-circle"
        cx={40}
        cy={-40}
        r="3"
      />
      <text
        className="m-ladder__timepoint-name"
        data-event="click"
        data-tip="MORTN Name"
        dominantBaseline="middle"
        textAnchor="middle"
        x="0"
        y={-40}
      >
        MORTN
      </text>
    </svg>
    <div
      className="__react_component_tooltip place-top type-dark"
      data-id="tooltip"
    />
  </div>,
  <div
    className="m-layover-box m-layover-box--bottom"
  />,
  <div
    className="m-incoming-box"
  />,
]
`;

exports[`routeLadder renders a route ladder with laying over vehicles 1`] = `
Array [
  <div
    className="m-route-ladder__header"
  >
    <button
      className="m-close-button"
      onClick={[Function]}
    >
      <span
        className=""
        dangerouslySetInnerHTML={
          Object {
            "__html": "SVG",
          }
        }
      />
    </button>
    <div
      className="m-route-ladder__route-name"
    >
      28
    </div>
  </div>,
  <div
    className="m-route-ladder__controls"
  >
    <button
      className="m-route-ladder__reverse"
      onClick={[Function]}
    >
      <span
        className="m-route-ladder__reverse-icon"
        dangerouslySetInnerHTML={
          Object {
            "__html": "SVG",
          }
        }
      />
      Reverse
    </button>
  </div>,
  <div
    className="m-layover-box m-layover-box--top"
  >
    <div
      className="m-layover-box__vehicle"
      onClick={[Function]}
    >
      <svg
        style={
          Object {
            "height": 27.72,
            "width": 26,
          }
        }
        viewBox="-13 -8.36 26 27.72"
      >
        <g
          className="m-vehicle-icon m-vehicle-icon--small on-time"
        >
          <rect
            className="m-vehicle-icon__label-background"
            height={11}
            rx={5.5}
            ry={5.5}
            width={26}
            x={-13}
            y={7.359999999999999}
          />
          <text
            className="m-vehicle-icon__label m-vehicle-icon__label--normal"
            dominantBaseline="central"
            textAnchor="middle"
            x="0"
            y={12.86}
          >
            2
          </text>
          <path
            className="m-vehicle-icon__triangle"
            d="m27.34 9.46 16.84 24.54a4.06 4.06 0 0 1 -1 5.64 4.11 4.11 0 0 1 -2.3.71h-33.72a4.06 4.06 0 0 1 -4.06-4.11 4 4 0 0 1 .72-2.24l16.84-24.54a4.05 4.05 0 0 1 5.64-1.05 4 4 0 0 1 1.04 1.05z"
            transform="scale(0.38) rotate(270) translate(-24,-22)"
          />
        </g>
      </svg>
    </div>
  </div>,
  <div
    className="m-ladder"
    style={
      Object {
        "width": 184,
      }
    }
  >
    <svg
      className="m-ladder__svg"
      height={0}
      viewBox="-92 -20 184 0"
      width={184}
    >
      <line
        className="m-ladder__line"
        x1={-40}
        x2={-40}
        y1="0"
        y2={-40}
      />
      <line
        className="m-ladder__line"
        x1={40}
        x2={40}
        y1="0"
        y2={-40}
      />
      <circle
        className="m-ladder__stop-circle"
        cx={-40}
        cy={-0}
        r="3"
      />
      <circle
        className="m-ladder__stop-circle"
        cx={40}
        cy={-0}
        r="3"
      />
      <text
        className="m-ladder__timepoint-name"
        data-event="click"
        data-tip="MATPN Name"
        dominantBaseline="middle"
        textAnchor="middle"
        x="0"
        y={-0}
      >
        MATPN
      </text>
      <circle
        className="m-ladder__stop-circle"
        cx={-40}
        cy={-20}
        r="3"
      />
      <circle
        className="m-ladder__stop-circle"
        cx={40}
        cy={-20}
        r="3"
      />
      <text
        className="m-ladder__timepoint-name"
        data-event="click"
        data-tip="WELLH Name"
        dominantBaseline="middle"
        textAnchor="middle"
        x="0"
        y={-20}
      >
        WELLH
      </text>
      <circle
        className="m-ladder__stop-circle"
        cx={-40}
        cy={-40}
        r="3"
      />
      <circle
        className="m-ladder__stop-circle"
        cx={40}
        cy={-40}
        r="3"
      />
      <text
        className="m-ladder__timepoint-name"
        data-event="click"
        data-tip="MORTN Name"
        dominantBaseline="middle"
        textAnchor="middle"
        x="0"
        y={-40}
      >
        MORTN
      </text>
    </svg>
    <div
      className="__react_component_tooltip place-top type-dark"
      data-id="tooltip"
    />
  </div>,
  <div
    className="m-layover-box m-layover-box--bottom"
  >
    <div
      className="m-layover-box__vehicle"
      onClick={[Function]}
    >
      <svg
        style={
          Object {
            "height": 27.72,
            "width": 26,
          }
        }
        viewBox="-13 -8.36 26 27.72"
      >
        <g
          className="m-vehicle-icon m-vehicle-icon--small on-time"
        >
          <rect
            className="m-vehicle-icon__label-background"
            height={11}
            rx={5.5}
            ry={5.5}
            width={26}
            x={-13}
            y={7.359999999999999}
          />
          <text
            className="m-vehicle-icon__label m-vehicle-icon__label--normal"
            dominantBaseline="central"
            textAnchor="middle"
            x="0"
            y={12.86}
          >
            1
          </text>
          <path
            className="m-vehicle-icon__triangle"
            d="m27.34 9.46 16.84 24.54a4.06 4.06 0 0 1 -1 5.64 4.11 4.11 0 0 1 -2.3.71h-33.72a4.06 4.06 0 0 1 -4.06-4.11 4 4 0 0 1 .72-2.24l16.84-24.54a4.05 4.05 0 0 1 5.64-1.05 4 4 0 0 1 1.04 1.05z"
            transform="scale(0.38) rotate(90) translate(-24,-22)"
          />
          <text
            className="m-vehicle-icon__variant"
            dominantBaseline="central"
            textAnchor="start"
            x={-5.6}
            y={0}
          >
            4
          </text>
        </g>
      </svg>
    </div>
  </div>,
  <div
    className="m-incoming-box"
  />,
]
`;

exports[`routeLadder renders a route ladder with vehicles 1`] = `
Array [
  <div
    className="m-route-ladder__header"
  >
    <button
      className="m-close-button"
      onClick={[Function]}
    >
      <span
        className=""
        dangerouslySetInnerHTML={
          Object {
            "__html": "SVG",
          }
        }
      />
    </button>
    <div
      className="m-route-ladder__route-name"
    >
      28
    </div>
  </div>,
  <div
    className="m-route-ladder__controls"
  >
    <button
      className="m-route-ladder__reverse"
      onClick={[Function]}
    >
      <span
        className="m-route-ladder__reverse-icon"
        dangerouslySetInnerHTML={
          Object {
            "__html": "SVG",
          }
        }
      />
      Reverse
    </button>
  </div>,
  <div
    className="m-layover-box m-layover-box--top"
  />,
  <div
    className="m-ladder"
    style={
      Object {
        "width": 232,
      }
    }
  >
    <svg
      className="m-ladder__svg"
      height={0}
      viewBox="-116 -20 232 0"
      width={232}
    >
      <line
        className="m-ladder__scheduled-line on-time"
        x1={-63}
        x2={-40}
        y1={-40}
        y2={-40}
      />
      <g>
        <g
          className="m-ladder__vehicle  "
          onClick={[Function]}
          transform="translate(87,-10)"
        >
          <g
            className="m-vehicle-icon m-vehicle-icon--medium on-time"
          >
            <rect
              className="m-vehicle-icon__label-background"
              height={11}
              rx={5.5}
              ry={5.5}
              width={26}
              x={-13}
              y={11.5}
            />
            <text
              className="m-vehicle-icon__label m-vehicle-icon__label--normal"
              dominantBaseline="central"
              textAnchor="middle"
              x="0"
              y={17}
            >
              1
            </text>
            <path
              className="m-vehicle-icon__triangle"
              d="m27.34 9.46 16.84 24.54a4.06 4.06 0 0 1 -1 5.64 4.11 4.11 0 0 1 -2.3.71h-33.72a4.06 4.06 0 0 1 -4.06-4.11 4 4 0 0 1 .72-2.24l16.84-24.54a4.05 4.05 0 0 1 5.64-1.05 4 4 0 0 1 1.04 1.05z"
              transform="scale(0.625) rotate(0) translate(-24,-22)"
            />
            <text
              className="m-vehicle-icon__variant"
              dominantBaseline="alphabetic"
              textAnchor="middle"
              x={0}
              y={8.5}
            >
              4
            </text>
          </g>
        </g>
      </g>
      <g>
        <g
          className="m-ladder__vehicle  "
          onClick={[Function]}
          transform="translate(63,-40)"
        >
          <g
            className="m-vehicle-icon m-vehicle-icon--medium ghost"
          >
            <rect
              className="m-vehicle-icon__label-background"
              height={11}
              rx={5.5}
              ry={5.5}
              width={26}
              x={-13}
              y={11.5}
            />
            <text
              className="m-vehicle-icon__label m-vehicle-icon__label--normal"
              dominantBaseline="central"
              textAnchor="middle"
              x="0"
              y={17}
            >
              0123
            </text>
            <g
              transform="scale(0.4375) translate(-24,-23)"
            >
              <path
                className="m-vehicle-icon__ghost-highlight"
                d="m43.79 19c0-9.68-8.79-17.49-19.59-17.49s-19.6 7.81-19.6 17.49v12.88 11a2 2 0 0 0 2.55 1.87l6.78-4.09 10.27 5.92 10.26-5.88 6.78 4.09a2 2 0 0 0 2.55-1.87z"
                stroke-join="round"
              />
              <path
                className="m-vehicle-icon__ghost-body"
                d="m43.79 19c0-9.68-8.79-17.49-19.59-17.49s-19.6 7.81-19.6 17.49v12.88 11a2 2 0 0 0 2.55 1.87l6.78-4.09 10.27 5.92 10.26-5.88 6.78 4.09a2 2 0 0 0 2.55-1.87z"
                stroke-join="round"
              />
              <ellipse
                className="m-vehicle-icon__ghost-eye"
                cx="19.73"
                cy="22.8"
                rx="3.11"
                ry="3.03"
              />
              <ellipse
                className="m-vehicle-icon__ghost-eye"
                cx="35.29"
                cy="22.8"
                rx="3.11"
                ry="3.03"
              />
            </g>
          </g>
        </g>
      </g>
      <g>
        <g
          className="m-ladder__vehicle  "
          onClick={[Function]}
          transform="translate(-63,-40)"
        >
          <g
            className="m-vehicle-icon m-vehicle-icon--medium on-time"
          >
            <rect
              className="m-vehicle-icon__label-background"
              height={11}
              rx={5.5}
              ry={5.5}
              width={26}
              x={-13}
              y={-22.5}
            />
            <text
              className="m-vehicle-icon__label m-vehicle-icon__label--normal"
              dominantBaseline="central"
              textAnchor="middle"
              x="0"
              y={-17}
            >
              2
            </text>
            <path
              className="m-vehicle-icon__triangle"
              d="m27.34 9.46 16.84 24.54a4.06 4.06 0 0 1 -1 5.64 4.11 4.11 0 0 1 -2.3.71h-33.72a4.06 4.06 0 0 1 -4.06-4.11 4 4 0 0 1 .72-2.24l16.84-24.54a4.05 4.05 0 0 1 5.64-1.05 4 4 0 0 1 1.04 1.05z"
              transform="scale(0.625) rotate(180) translate(-24,-22)"
            />
          </g>
        </g>
      </g>
      <line
        className="m-ladder__line"
        x1={-40}
        x2={-40}
        y1="0"
        y2={-40}
      />
      <line
        className="m-ladder__line"
        x1={40}
        x2={40}
        y1="0"
        y2={-40}
      />
      <circle
        className="m-ladder__stop-circle"
        cx={-40}
        cy={-0}
        r="3"
      />
      <circle
        className="m-ladder__stop-circle"
        cx={40}
        cy={-0}
        r="3"
      />
      <text
        className="m-ladder__timepoint-name"
        data-event="click"
        data-tip="MATPN Name"
        dominantBaseline="middle"
        textAnchor="middle"
        x="0"
        y={-0}
      >
        MATPN
      </text>
      <circle
        className="m-ladder__stop-circle"
        cx={-40}
        cy={-20}
        r="3"
      />
      <circle
        className="m-ladder__stop-circle"
        cx={40}
        cy={-20}
        r="3"
      />
      <text
        className="m-ladder__timepoint-name"
        data-event="click"
        data-tip="WELLH Name"
        dominantBaseline="middle"
        textAnchor="middle"
        x="0"
        y={-20}
      >
        WELLH
      </text>
      <circle
        className="m-ladder__stop-circle"
        cx={-40}
        cy={-40}
        r="3"
      />
      <circle
        className="m-ladder__stop-circle"
        cx={40}
        cy={-40}
        r="3"
      />
      <text
        className="m-ladder__timepoint-name"
        data-event="click"
        data-tip="MORTN Name"
        dominantBaseline="middle"
        textAnchor="middle"
        x="0"
        y={-40}
      >
        MORTN
      </text>
    </svg>
    <div
      className="__react_component_tooltip place-top type-dark"
      data-id="tooltip"
    />
  </div>,
  <div
    className="m-layover-box m-layover-box--bottom"
  />,
  <div
    className="m-incoming-box"
  />,
]
`;

exports[`routeLadder renders a route ladder with vehicles in the incoming box 1`] = `
Array [
  <div
    className="m-route-ladder__header"
  >
    <button
      className="m-close-button"
      onClick={[Function]}
    >
      <span
        className=""
        dangerouslySetInnerHTML={
          Object {
            "__html": "SVG",
          }
        }
      />
    </button>
    <div
      className="m-route-ladder__route-name"
    >
      28
    </div>
  </div>,
  <div
    className="m-route-ladder__controls"
  >
    <button
      className="m-route-ladder__reverse"
      onClick={[Function]}
    >
      <span
        className="m-route-ladder__reverse-icon"
        dangerouslySetInnerHTML={
          Object {
            "__html": "SVG",
          }
        }
      />
      Reverse
    </button>
  </div>,
  <div
    className="m-layover-box m-layover-box--top"
  />,
  <div
    className="m-ladder"
    style={
      Object {
        "width": 184,
      }
    }
  >
    <svg
      className="m-ladder__svg"
      height={0}
      viewBox="-92 -20 184 0"
      width={184}
    >
      <line
        className="m-ladder__line"
        x1={-40}
        x2={-40}
        y1="0"
        y2={-40}
      />
      <line
        className="m-ladder__line"
        x1={40}
        x2={40}
        y1="0"
        y2={-40}
      />
      <circle
        className="m-ladder__stop-circle"
        cx={-40}
        cy={-0}
        r="3"
      />
      <circle
        className="m-ladder__stop-circle"
        cx={40}
        cy={-0}
        r="3"
      />
      <text
        className="m-ladder__timepoint-name"
        data-event="click"
        data-tip="MATPN Name"
        dominantBaseline="middle"
        textAnchor="middle"
        x="0"
        y={-0}
      >
        MATPN
      </text>
      <circle
        className="m-ladder__stop-circle"
        cx={-40}
        cy={-20}
        r="3"
      />
      <circle
        className="m-ladder__stop-circle"
        cx={40}
        cy={-20}
        r="3"
      />
      <text
        className="m-ladder__timepoint-name"
        data-event="click"
        data-tip="WELLH Name"
        dominantBaseline="middle"
        textAnchor="middle"
        x="0"
        y={-20}
      >
        WELLH
      </text>
      <circle
        className="m-ladder__stop-circle"
        cx={-40}
        cy={-40}
        r="3"
      />
      <circle
        className="m-ladder__stop-circle"
        cx={40}
        cy={-40}
        r="3"
      />
      <text
        className="m-ladder__timepoint-name"
        data-event="click"
        data-tip="MORTN Name"
        dominantBaseline="middle"
        textAnchor="middle"
        x="0"
        y={-40}
      >
        MORTN
      </text>
    </svg>
    <div
      className="__react_component_tooltip place-top type-dark"
      data-id="tooltip"
    />
  </div>,
  <div
    className="m-layover-box m-layover-box--bottom"
  />,
  <div
    className="m-incoming-box"
  >
    <button
      className="m-incoming-box__vehicle "
      onClick={[Function]}
    >
      <svg
        style={
          Object {
            "height": 15.2,
            "width": 16.72,
          }
        }
        viewBox="-8.36 -7.6 16.72 15.2"
      >
        <g
          className="m-vehicle-icon m-vehicle-icon--small on-time"
        >
          <path
            className="m-vehicle-icon__triangle"
            d="m27.34 9.46 16.84 24.54a4.06 4.06 0 0 1 -1 5.64 4.11 4.11 0 0 1 -2.3.71h-33.72a4.06 4.06 0 0 1 -4.06-4.11 4 4 0 0 1 .72-2.24l16.84-24.54a4.05 4.05 0 0 1 5.64-1.05 4 4 0 0 1 1.04 1.05z"
            transform="scale(0.38) rotate(0) translate(-24,-22)"
          />
          <text
            className="m-vehicle-icon__variant"
            dominantBaseline="alphabetic"
            textAnchor="middle"
            x={0}
            y={5.6}
          >
            4
          </text>
        </g>
      </svg>
      <div
        className="m-incoming-box__vehicle-label"
      >
        1
      </div>
    </button>
    <button
      className="m-incoming-box__vehicle "
      onClick={[Function]}
    >
      <svg
        style={
          Object {
            "height": 15.2,
            "width": 16.72,
          }
        }
        viewBox="-8.36 -7.6 16.72 15.2"
      >
        <g
          className="m-vehicle-icon m-vehicle-icon--small on-time"
        >
          <path
            className="m-vehicle-icon__triangle"
            d="m27.34 9.46 16.84 24.54a4.06 4.06 0 0 1 -1 5.64 4.11 4.11 0 0 1 -2.3.71h-33.72a4.06 4.06 0 0 1 -4.06-4.11 4 4 0 0 1 .72-2.24l16.84-24.54a4.05 4.05 0 0 1 5.64-1.05 4 4 0 0 1 1.04 1.05z"
            transform="scale(0.38) rotate(180) translate(-24,-22)"
          />
        </g>
      </svg>
      <div
        className="m-incoming-box__vehicle-label"
      >
        2
      </div>
    </button>
  </div>,
]
`;
